### ISSUE TITLE
AsciiMath - cannot add functions to symbols table

### ISSUE THREAD

**Issue by bhelzer** [mathjax/MathJax#1296]
I used to be able to add functions to the input table by calling:

``` javascript
AM = MathJax.InputJax.AsciiMath.AM;
AM.symbols.push({
  func: true,
  input: 'asin',
  output: 'asin',
  tag: 'mi',
  tex: null,
  type: AM.TOKEN.UNARY
 })
```

It appears that the new library initializes AM.symbols to the symbol table, but then makes copies of it every time somebody modifies it. If I am the first user, I can make the changes I want, but I don't have any control over that. 
I could use the `define` method, but it doesn't allow we to set the tag, or func attributes, so the parsing is incorrect. Can we make one of to changes:
1. (Preferred) Enhance define to allow we to set tag and func, maybe let me pass in the object I want to set. It would be easy enough to detect the difference between 2 strings or an object. 
2. Or, keep AM.symbols up to date so I can modify it.

**Comment by dpvc** [comment]
You don't give the details about where you run the code you provide, but it should be enclosed in 

```
<script type="text/x-mathjax-config">
MathJax.Hub.Register.StartupHook("AsciiMath Jax Config",function () {
  ...
}
</script>
```

and that should be *before* the the script the loads `MathJax.js` itself.  See this [documentation](https://github.com/mathjax/mathjax-docs/wiki/Expanding-AsciiMath-syntax) for an example.

The AsciiMath input jax uses the `AsciiMathML.js` nearly verbatim, so things like the definition of `define` are not something that we would change, so enhancements should be requested at the [AsciiMath](https://github.com/asciimath/asciimathml/) repository.

It's true, however, that AsciiMath does create a new copy of the `AMsymbols` array when new commands are added (and that occurs during initialization), so the `AM.symbols` value is out of date and modifying it will have not effect, as you point out.  Perhaps `AM.symbols` should be a function that returns the current AMsymbols array instead.

**Comment by dpvc** [comment]
Making `AM.symbols` use a getter to return the current `AMsymbols` array would make it backward compatible, so that is probably the better approach.  Then you could do

```
AM = MathJax.InputJax.AsciiMath.AM;
AM.symbols.push({
  func: true,
  input: 'asin',
  output: 'asin',
  tag: 'mi',
  tex: null,
  type: AM.TOKEN.UNARY
 });
AM.refreshSymbols();
```

at any time.

**Comment by bhelzer** [comment]
The program in fact does register the startup hook, which works perfectly when I load MathJax dynamically. The code I am working on is in a library, and the application that reported the error loads MathJax on startup, before I have a chance to register the hook. By the time I register it, it is too late, the array has already been copied.

**Comment by pkra** [comment]
Was this filed upstream? cc @drlippman

**Comment by dpvc** [comment]
It is only partly an upstream issue.  It is more about the glue between AsciiMath and MathJax.  It could be fixed by AsciiMath changing how the symbol array is handled, but we can also fix it in MathJax itself.

**Comment by drlippman** [comment]
Would it work to create a new function for adding symbols, that allows the full symbol object to be provided as input?  That seems like it'd be easy enough to add to AsciiMath.  

If you have a suggestion for a bigger change to how the symbol array is handled, that'd be interesting to hear too.

**Comment by dpvc** [comment]
@drlippman:

> If you have a suggestion for a bigger change to how the symbol array is handled, that'd be interesting to hear too.

The main problem is that the array pointed to by the `AMsymbols` variable can change, and so the glue that allows that symbol to be viewed from outside its sandbox in MathJax keeps pointing to the old array while the internal variable points to a new one.  (If MathJax used a getter instead of a variable, that would solve the problem with no change to AsciiMath, but getters and setters aren't implements in IE prior to 9, so that would break MathJax's AsciiMath input jax in IE8 and below.)

Instead, if you were to modify the existing array rather than create a new one, that would help.  For example, [lines 190 and 191](https://github.com/asciimath/asciimathml/blob/master/ASCIIMathML.js#L190-L191) change `AMsymbols` using `concat()` which forms a new array:

```
  AMsymbols = AMsymbols.concat([{input:oldstr, tag:"mo", output:newstr, 
                                 tex:null, ttype:DEFINITION}]);
```

while they could use `push()` instead, which does not.

```
  AMsymbols.push({input:oldstr, tag:"mo", output:newstr, tex:null, ttype:DEFINITION});
```
(note the removal of square brackets from the argument as well).

Similarly for [lines 492-493](https://github.com/asciimath/asciimathml/blob/master/ASCIIMathML.js#L492-L493):

```
  AMsymbols = AMsymbols.concat([{input:oldstr, tag:"mo", output:newstr, 
                                 tex:null, ttype:DEFINITION}]);
```
could be
```
  AMsymbols.push({input:oldstr, tag:"mo", output:newstr, tex:null, ttype:DEFINITION});
```

while [line 481](https://github.com/asciimath/asciimathml/blob/master/ASCIIMathML.js#L481)

```
 AMsymbols = AMsymbols.concat(texsymbols);
```
could be
```
AMsymbols.push.apply(AMsymbols,texsymbols);
```
(or you could just push the symbols onto AMsymbols within the loop at line 447 rather than adding them to a temporary array first; even if you don't the assignment on 447 could also be done via `push()` instead).

From my quick scan of the file, those are the three places where the `AMsymbols` variable changes the array.  If those where altered as suggested above, this issue would go away for MathJax.

**Comment by drlippman** [comment]
Makes sense to me.  I'll work on it.

I'm presuming we'd still need a new function for adding an arbitrary symbol, since even if the same array is used, pushing a new symbol onto it after the array is initialized still requires a resort of the array.

**Comment by drlippman** [comment]
Made the concat/push changes upstream in asciimath/asciimathml@f649ba49f639b7e5322d6552193226c03e88ba7e

I also went ahead and added a new `newsymbol` command that will take a symbol object and push it onto the symbols array and refresh the array.

**Comment by dpvc** [comment]
Thanks!  I'll have to try it out, but the new `newsymbol()` command should resolve any issues from the MathJax end.

**Comment by beeplin** [comment]
This issue should be closed since v2.7.0 has added `newsymbol()` and make `newcommand` use `push` instead of `concat`. Now it works fine like this: 
```js
    MathJax.Hub.Register.StartupHook('AsciiMath Jax Ready', function () {
      AM = MathJax.InputJax.AsciiMath.AM
      AM.newsymbol({ input: '>=', tag: 'mo', output: '\u2A7E', tex: 'geqslant', ttype: AM.TOKEN.CONST })
      AM.newsymbol({ input: '<=', tag: 'mo', output: '\u2A7D', tex: 'leqslant', ttype: AM.TOKEN.CONST })
      AM.newsymbol({ input: 'comp', tag: 'mo', output: '\u2201', tex: 'complement', ttype: AM.TOKEN.CONST })
      AM.newsymbol({ input: 'sim', tag: 'mo', output: '\u223C', tex: null, ttype: AM.TOKEN.CONST })
      AM.newsymbol({ input: 'sub', tag: 'mo', output: '\u2ACB', tex: 'subsetneqq', ttype: AM.TOKEN.CONST })
      AM.newsymbol({ input: 'sup', tag: 'mo', output: '\u2ACC', tex: 'supsetneqq', ttype: AM.TOKEN.CONST })
      AM.newsymbol({ input: "//", tag: "mo", output: "\u2225", tex: null, ttype: AM.TOKEN.CONST })
      AM.newsymbol({ input: "||", tag: "mo", output: "\u2225", tex: null, ttype: AM.TOKEN.CONST })
      AM.newsymbol({ input: "!//", tag: "mo", output: "\u2226", tex: null, ttype: AM.TOKEN.CONST })
      AM.newsymbol({ input: "!||", tag: "mo", output: "\u2226", tex: null, ttype: AM.TOKEN.CONST })
    })
```

**Comment by pkra** [comment]
Thanks for pointing out that we forgot to close this here. It was only mentioned in the release notes.
